ts/src/connections/utils.ts

```diff
@@ -19,3 +22,92 @@ export class List {
         return map;
     }
 }
+
+/**
+ * Resolves the promise after {@link milliseconds} have passed, or reject if {@link cancellation} is canceled.
+ */
+export function delay(milliseconds: number, cancellation?: CancellationToken): Promise<void> {
+    return new Promise<void>((resolve, reject) => {
+        let cancellationDisposable: Disposable | undefined;
+        let timeout: NodeJS.Timeout | undefined;
+
+        if (cancellation) {
+            if (cancellation.isCancellationRequested) {
+                reject(new CancellationError());
+                return;
+            }
+
+            cancellationDisposable = cancellation.onCancellationRequested(() => {
+                if (timeout) {
+                    clearTimeout(timeout);
+                }
+
+                cancellationDisposable?.dispose();
+                reject(new CancellationError());
+            });
+        }
+
+        timeout = setTimeout(() => {
+            cancellationDisposable?.dispose();
+            resolve();
+        }, milliseconds);
+    });
+}
+
+/**
+ * Checks if {@link e} is {@link CancellationError} and {@link cancellation} has been requested.
+ */
+export function isCancellation(e: any, cancellation?: CancellationToken): e is CancellationError {
+    return (
+        e &&
+        typeof e === 'object' &&
+        e instanceof CancellationError &&
```

Comment:
nit: `instanceof` makes both of the prior checks unnecessary. A null/undefined or non-object value will always fail an `instanceof` check.